Support positional matching

AstraLuma opened the pull request:
New for 3.10.

AstraLuma committed:
Support positional matching

New for 3.10

AstraLuma commented:
Not sure how to handle testing for this, given that the preferred tests won't even compile on <3.10.

nicoonoclaste commented:
@AstraLuma I'd suggest having a `test_py3.10.py` or such, and [selectively ignoring it in `conftest.py`](https://docs.pytest.org/en/6.2.x/example/pythoncollection.html#customizing-test-collection) based on Python's version number.

AstraLuma committed:
Add tests

AstraLuma committed:
Add docs

AstraLuma committed:
Fix a typo

AstraLuma committed:
Apparently has to be a top-level conftest.py

AstraLuma commented:
ok, now i officially have no idea why pytest isn't collecting properly.

AstraLuma committed:
Hopefully this variant works?

AstraLuma committed:
Update python images

AstraLuma committed:
Ignore B011 for tests

B011: B011 Do not call assert False since python -O removes these calls. Instead callers should raise AssertionError().

AstraLuma committed:
Fix linting

AstraLuma committed:
Merge branch 'canon' into add-matching

AstraLuma committed:
Exclude pattern matching from mypy

https://github.com/python/mypy/pull/10191

AstraLuma committed:
moar lint

AstraLuma commented:
@nbraud @pathunstrom I think this is ready for review.

pathunstrom reviewed (approved):
I like the methodology for testing forward looking features when they don't break syntax.

nicoonoclaste committed:
.common.sh: Add a `die` function (minor refactor)

nicoonoclaste committed:
.common.sh: Extract Python version, assert Py3

nicoonoclaste committed:
lint.sh: Do not lint Py3.10+ tests on older Python

nicoonoclaste reviewed (approved):
LGTM, except it broke `lint.sh` on Python versions older than 3.10 (not unlikely on contributors' dev environment)

I wrote a fix for that, please review it and if it's fine by you then `bors r=nbraud,pathunstrom` ?

nicoonoclaste commented on ppb_vector/__init__.py:
```diff
@@ -64,6 +64,8 @@ class Vector:
     x: float
     y: float
 
+    __match_args__ = ('x', 'y')
```
Might be nice to add a comment pointing to the right documentation (in the PEP or otherwise) for that dunder ^^

AstraLuma commented:
Oh, I just fixed it by updating the containers. But I guess that helps with development environments.

AstraLuma commented:
bors r=nbraud,pathunstrom
